### DISCUSSION TITLE
How to properly use indexes to speed up queries?

### DISCUSSION THREAD

**Discussion by FelipeLagare** [duckdb/duckdb#4006] [Q&A]
I currently have a scenario where it is needed to constantly (~1 sec for dashboard purposes) query distinct values from 5 or more columns from a database. The issue is, that these 5 columns can have up to half a million distinct values each. 
I'm using aggregates, listing on the columns and I've tried querying each column separately using index but it seems to not make a big difference. 

What would be a better solution for this in duckdb databases?

**Comment by Mytherin** [comment]
Indexes are generally not used to speed up aggregations. Could you share the queries you are running?

**Reply by FelipeLagare** [reply]
Sorry, let me rephrase it: in order to speed up the db connection time on python I've created a new db using aggregations from other db tables so it reduces the row count.

Aside from that, on the new db, I need to use the following query:

df = con.execute(
"""
WITH tbl AS (
  SELECT s1, s2, s3, s4, s5 
  FROM table
  WHERE n = '150'
),
c1 AS (SELECT 's1' as col, s1 as value FROM tbl),
c2 AS (SELECT 's2' as col, s2 as value FROM tbl),
c3 AS (SELECT 's3' as col, s3 as value FROM tbl),
c4 AS (SELECT 's4' as col, s4 as value FROM tbl),
c5 AS (SELECT 's5' as col, s5 as value FROM tbl)
select col, list(value) from c1
group by col
union all
select col, list(value) from c2
group by col
union all
select col, list(value) from c3
group by col
union all
select col, list(value) from c4
group by col
union all
select col, list(value) from c5
group by col
"""
).fetchdf()

The indexes I've tried were for each one of the s1, s2, s3, s4, s5 columns before running the query above, they were simply created like:
CREATE INDEX idx ON table (s1)

**Reply by FelipeLagare** [reply]
Summarizing, I'm trying to speed a SELECT DISTINCT by using an index.  I tried creating one in a single column, but it didn't improve performance at all.  Are indexes the right tool to use for this?

**Reply by Mytherin** [reply]
Indexes are only used if they can drastically reduce the number of rows you need to scan. In this case,  the only filter is on the column "n", so that is the only index that makes sense. An index on any other column will not have any effect. 

```sql
SELECT s1, s2, s3, s4, s5
FROM table
WHERE n = '150'
```

Your query as it is is a bit confusing - are you grouping by a constant value, and then generating a giant list of values of each column separately? Are all the unions necessary, i.e. can't you just do this in one pass over the table?

**Reply by FelipeLagare** [reply]
The purpose of the query is, given a set of filtering conditions (in this case n = '150'), we want the distinct values of a set of columns after applying that filtering.  Those values in reality represent options that can be selected from a dropdown list. So, after selecting one value (for the dropdown "n"), we want to re-populate each dropdown with new options.  It could be possible to do it in a single pass, let me give it some thought.

**Reply by Mytherin** [reply]
Is there any reason to use the `LIST` aggregate here, rather than just doing a `GROUP BY` and reading all of the groups?
